### ISSUE TITLE
Wrong argument validation in `DeviceMesh._unflatten`

### ISSUE THREAD

**Issue by antony-frolov** [pytorch/pytorch#176443]
### 🐛 Describe the bug

Hi! It seems like the validation conditional is wrong in `DeviceMesh._unflatten`. It checks if `dim` is in `self.mesh_dim_names`, but the error message below says that it's `dim` is not in `self.mesh_dim_names`. I suppose the expression in the if statement is wrong and it's supposed to be `elif isinstance(dim, str) and dim not in not_none(self.mesh_dim_names)`

https://github.com/pytorch/pytorch/blob/d7063960c58fcd47cea2aeb6eff41e6234b8d2a1/torch/distributed/device_mesh.py#L1378

### Versions

PyTorch version: 2.10.0a0+git449b176
Is debug build: False
CUDA used to build PyTorch: 13.0
ROCM used to build PyTorch: N/A

OS: Ubuntu 24.04.3 LTS (x86_64)
GCC version: (Ubuntu 13.3.0-6ubuntu2~24.04) 13.3.0
Clang version: Could not collect
CMake version: version 3.28.3
Libc version: glibc-2.39

Python version: 3.12.12 (main, Dec  2 2025, 19:47:35) [Clang 21.1.4 ] (64-bit runtime)
Python platform: Linux-5.4.210-39.1-x86_64-with-glibc2.39
Is CUDA available: True
CUDA runtime version: Could not collect
CUDA_MODULE_LOADING set to: 
GPU models and configuration: 
GPU 0: NVIDIA A100-SXM4-40GB
GPU 1: NVIDIA A100-SXM4-40GB
GPU 2: NVIDIA A100-SXM4-40GB
GPU 3: NVIDIA A100-SXM4-40GB
GPU 4: NVIDIA A100-SXM4-40GB
GPU 5: NVIDIA A100-SXM4-40GB
GPU 6: NVIDIA A100-SXM4-40GB
GPU 7: NVIDIA A100-SXM4-40GB

Nvidia driver version: 535.129.03
cuDNN version: Probably one of the following:
/usr/lib/x86_64-linux-gnu/libcudnn.so.9.13.0
/usr/lib/x86_64-linux-gnu/libcudnn_adv.so.9.13.0
/usr/lib/x86_64-linux-gnu/libcudnn_cnn.so.9.13.0
/usr/lib/x86_64-linux-gnu/libcudnn_engines_precompiled.so.9.13.0
/usr/lib/x86_64-linux-gnu/libcudnn_engines_runtime_compiled.so.9.13.0
/usr/lib/x86_64-linux-gnu/libcudnn_graph.so.9.13.0
/usr/lib/x86_64-linux-gnu/libcudnn_heuristic.so.9.13.0
/usr/lib/x86_64-linux-gnu/libcudnn_ops.so.9.13.0
Is XPU available: False
HIP runtime version: N/A
MIOpen runtime version: N/A
Is XNNPACK available: True
Caching allocator config: N/A

CPU:
Architecture:                    x86_64
CPU op-mode(s):                  32-bit, 64-bit
Address sizes:                   48 bits physical, 48 bits virtual
Byte Order:                      Little Endian
CPU(s):                          120
On-line CPU(s) list:             0-119
Vendor ID:                       AuthenticAMD
Model name:                      AMD EPYC 7662 64-Core Processor
CPU family:                      23
Model:                           49
Thread(s) per core:              1
Core(s) per socket:              1
Socket(s):                       120
Stepping:                        0
BogoMIPS:                        3992.09
Flags:                           fpu vme de pse tsc msr pae mce cx8 apic sep mtrr pge mca cmov pat pse36 clflush mmx fxsr sse sse2 syscall nx mmxext fxsr_opt pdpe1gb rdtscp lm constant_tsc rep_good nopl nonstop_tsc cpuid extd_apicid pni pclmulqdq ssse3 fma cx16 sse4_1 sse4_2 x2apic movbe popcnt tsc_deadline_timer aes xsave avx f16c rdrand hypervisor lahf_lm cmp_legacy svm cr8_legacy abm sse4a misalignsse 3dnowprefetch osvw perfctr_core ssbd ibrs ibpb stibp vmmcall fsgsbase tsc_adjust bmi1 avx2 smep bmi2 rdseed adx smap clflushopt clwb sha_ni xsaveopt xsavec xgetbv1 clzero xsaveerptr wbnoinvd arat npt nrip_save umip rdpid arch_capabilities
Virtualization:                  AMD-V
L1d cache:                       7.5 MiB (120 instances)
L1i cache:                       7.5 MiB (120 instances)
L2 cache:                        60 MiB (120 instances)
L3 cache:                        1.9 GiB (120 instances)
NUMA node(s):                    4
NUMA node0 CPU(s):               0-29
NUMA node1 CPU(s):               30-59
NUMA node2 CPU(s):               60-89
NUMA node3 CPU(s):               90-119
Vulnerability Itlb multihit:     Not affected
Vulnerability L1tf:              Not affected
Vulnerability Mds:               Not affected
Vulnerability Meltdown:          Not affected
Vulnerability Mmio stale data:   Not affected
Vulnerability Spec store bypass: Mitigation; Speculative Store Bypass disabled via prctl and seccomp
Vulnerability Spectre v1:        Mitigation; usercopy/swapgs barriers and __user pointer sanitization
Vulnerability Spectre v2:        Mitigation; Retpolines, IBPB conditional, IBRS_FW, STIBP disabled, RSB filling, PBRSB-eIBRS Not affected
Vulnerability Srbds:             Not affected
Vulnerability Tsx async abort:   Not affected

Versions of relevant libraries:
[pip3] flake8==7.3.0
[pip3] mypy==1.16.1
[pip3] mypy_extensions==1.1.0
[pip3] numpy==2.3.3
[pip3] onnx==1.20.0
[pip3] onnx-ir==0.1.12
[pip3] onnxscript==0.5.6
[pip3] torch==2.10.0a0+git449b176
[pip3] torch_c_dlpack_ext==0.1.5
[pip3] torchaudio==2.9.0+eaa9e4e
[pip3] torchdata==0.12.0.dev20260202+cpu
[pip3] torchvision==0.24.0+7a9db90
[pip3] triton==3.6.0+git8fedd49b
[conda] Could not collect

cc @awgu @wanchaol @fegin @fduwjj @wz337 @wconstab @d4l3k @pragupta @msaroufim @dcci @aditvenk @xmfan

**Comment by RohitRathore1** [comment]
cc @fduwjj — it seems is was introduced in your PR #161224.

using `_unflatten` with a valid string dim is completely broken. Any call like `mesh._unflatten("world", ...)` will incorrectly raise `ValueError`, even though "world" is a valid dim name. This blocks the string-based API for `_unflatten` entirely. Also, invalid string dims bypass validation and crash later at `.index(dim)` on with an unhelpful error, instead of the clear error message intended by the validation.

```
import torch
import torch.distributed as dist
from torch.distributed.device_mesh import init_device_mesh

dist.init_process_group(backend="nccl")

mesh_1d = init_device_mesh("cuda", (8,), mesh_dim_names=("world",))

# BUG: This should work but raises ValueError
try:
    mesh_3d = mesh_1d._unflatten("world", (2, 2, 2), ("dp", "cp", "tp"))
    print(f"[Rank {dist.get_rank()}] String dim succeeded: {mesh_3d.mesh_dim_names}")
except ValueError as e:
    print(f"[Rank {dist.get_rank()}] String dim FAILED: {e}")

# Integer dim works fine
try:
    mesh_3d = mesh_1d._unflatten(0, (2, 2, 2), ("dp", "cp", "tp"))
    print(f"[Rank {dist.get_rank()}] Integer dim succeeded: {mesh_3d.mesh_dim_names}")
except ValueError as e:
    print(f"[Rank {dist.get_rank()}] Integer dim FAILED: {e}")

dist.destroy_process_group()
```
